[AIRFLOW-2300] Add S3 Select functionarity to S3ToHiveTransfer

sekikn opened the pull request:
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2300
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

To improve efficiency and usability, this PR adds
S3 Select functionarity to S3ToHiveTransfer.
It also contains some minor fixes for documents
and comments.


### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added test_execute_with_select_expression to tests.operators.s3_to_hive_operator:S3ToHiveTransferTest


### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.


### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

artwr commented on airflow/hooks/S3_hook.py:
```diff
@@ -194,9 +194,9 @@ def select_key(self, key, bucket_name=None,
+        :type input_serialization: dict
         :param output_serialization: S3 Select output data serialization format
-        :type output_serialization: str
+        :type output_serialization: dict
```
I know we are not consistent about this a info about the return type would be helpful here.

artwr commented on airflow/operators/s3_to_hive_operator.py:
```diff
@@ -146,16 +150,42 @@ def execute(self, context):
                     "The key {0} does not exists".format(self.s3_key))
             s3_key_object = self.s3.get_key(self.s3_key)
+
+        if self.select_expression is not None and self.input_compressed\
```
A couple of comments: 
- This is more of a nit, but I am more of a fan of using parentheses for line continuation rather than the backslash.
- `self.select_expression is not None` is used twice, time to maybe add a local var `has_select_expression = self.select_expression is not None`, or just use `if self.select_expression`, as none of the false-y values are valid select_expressions anyway. 
- Similarly move the test below `root, file_ext = os.path.splitext(s3_key_object.key)` and get the file extension parsed for free :)

artwr reviewed (commented):
Some minor changes to clean s3_to_hive_operator.py up but otherwise looks good.

sekikn committed:
[AIRFLOW-2300] Add S3 Select functionarity to S3ToHiveTransfer

To improve efficiency and usability, this PR adds
S3 Select functionarity to S3ToHiveTransfer.
It also contains some minor fixes for documents
and comments.

Fokko reviewed (approved):
LGTM
